File: csharp/Dson/src/DsonType.cs
﻿#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// Dson数据类型枚举
/// </summary>
public enum DsonType
{
    /** 对象的结束标识 */
    EndOfObject = 0,

    Int32 = 1,
    Int64 = 2,
    Float = 3,
    Double = 4,
    Boolean = 5,
    String = 6,
    Null = 7,

    /// <summary>
    /// 二进制字节数组
    /// 我们为二进制字节数组也提供了一个扩展子类型，实现一定程度上的自解释
    /// </summary>
    Binary = 8,

    /// <summary>
    /// Int32的扩展类型
    /// 基本类型的int无法直接表达其使用目的，需要额外的类型支持；
    /// 通过扩展类型，可避免破坏业务代码，可避免用户自行封装
    /// </summary>
    ExtInt32 = 9,
    /// <summary>
    /// Int64的扩展类型
    /// </summary>
    ExtInt64 = 10,
    /// <summary>
    /// Double的扩展类型
    /// </summary>
    ExtDouble = 11,
    /// <summary>
    /// String的扩展类型
    /// </summary>
    ExtString = 12,

    /// <summary>
    /// 对象引用
    /// </summary>
    Reference = 13,
    /// <summary>
    /// 时间戳
    /// </summary>
    Timestamp = 14,

    /// <summary>
    /// 对象头信息，与Object类型编码格式类似
    /// 但header不可以再直接嵌入header
    /// </summary>
    Header = 29,
    /// <summary>
    /// 数组(v,v,v...)
    /// </summary>
    Array = 30,
    /// <summary>
    /// 普通对象(k,v,k,v...)
    /// </summary>
    Object = 31,
}

/// <summary>
/// DsonType的工具类
/// </summary>
public static class DsonTypes
{
    private static readonly DsonType[] LookUp;

    /** 用于表示无效DsonType的共享值 */
    public static readonly DsonType Invalid = (DsonType)(-1);

    static DsonTypes() {
        LookUp = new DsonType[(int)DsonType.Object + 1];
        foreach (var dsonType in Enum.GetValues<DsonType>()) {
            LookUp[(int)dsonType] = dsonType;
        }
    }

    /** DsonType是否表示Dson的Number */
    public static bool IsNumber(this DsonType dsonType) {
        return dsonType switch
        {
            DsonType.Int32 => true,
            DsonType.Int64 => true,
            DsonType.Float => true,
            DsonType.Double => true,
            _ => false
        };
    }

    /** DsonType关联的值在二进制编码时是否包含WireType */
    public static bool HasWireType(this DsonType dsonType) {
        return dsonType switch
        {
            DsonType.Int32 => true,
            DsonType.Int64 => true,
            DsonType.ExtInt32 => true,
            DsonType.ExtInt64 => true,
            _ => false
        };
    }

    /** DsonType是否表示容器类型；header不属于普通意义上的容器 */
    public static bool IsContainer(this DsonType dsonType) {
        return dsonType == DsonType.Object || dsonType == DsonType.Array;
    }

    /** DsonType是否是容器类型或Header */
    public static bool IsContainerOrHeader(this DsonType dsonType) {
        return dsonType == DsonType.Object || dsonType == DsonType.Array || dsonType == DsonType.Header;
    }

    /** Dson是否是KV结构 */
    public static bool IsObjectLike(this DsonType dsonType) {
        return dsonType == DsonType.Object || dsonType == DsonType.Header;
    }

    /** 通过Number获取对应的枚举 */
    public static DsonType ForNumber(int number) {
        return LookUp[number];
    }
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonType.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

/**
 * Dson数据类型
 *
 * @author wjybxx
 * date - 2023/4/19
 */
public enum DsonType {

    /** 对象的结束标识 */
    END_OF_OBJECT(0),

    INT32(1),
    INT64(2),
    FLOAT(3),
    DOUBLE(4),
    BOOLEAN(5),
    STRING(6),
    NULL(7),

    /**
     * 二进制字节数组
     * 我们为二进制字节数组也提供了一个扩展子类型，实现一定程度上的自解释
     */
    BINARY(8),
    /**
     * Int32的扩展类型
     * 基本类型的int无法直接表达其使用目的，需要额外的类型支持；
     * 通过扩展类型，可避免破坏业务代码，可避免用户自行封装
     */
    EXT_INT32(9),
    /**
     * Int64的扩展类型
     */
    EXT_INT64(10),
    /**
     * Double的扩展类型
     */
    EXT_DOUBLE(11),
    /**
     * String的扩展类型
     * 基本类型的String无法直接表达其使用目的，需要额外的类型支持；
     * 通过扩展类型，可避免破坏业务代码，可避免用户自行封装
     */
    EXT_STRING(12),
    /**
     * 对象引用
     */
    REFERENCE(13),
    /**
     * 时间戳
     */
    TIMESTAMP(14),

    /**
     * 对象头信息，与Object类型编码格式类似
     * 但header不可以再直接嵌入header
     */
    HEADER(29),
    /**
     * 数组(v,v,v...)
     */
    ARRAY(30),
    /**
     * 普通对象(k,v,k,v...)
     */
    OBJECT(31),
    ;
    private final int number;

    DsonType(int number) {
        assert number >= 0 && number <= Dsons.DSON_TYPE_MAX_VALUE;
        this.number = number;
    }

    private static final DsonType[] LOOK_UP = new DsonType[Dsons.DSON_TYPE_MAX_VALUE + 1];

    static {
        for (DsonType type : values()) {
            LOOK_UP[type.getNumber()] = type;
        }
    }

    public int getNumber() {
        return number;
    }

    public boolean isPrimitive() {
        return number >= 1 && number <= 6;
    }

    public boolean isNumber() {
        return switch (this) {
            case INT32, INT64, FLOAT, DOUBLE -> true;
            default -> false;
        };
    }

    /** {@link WireType} */
    public boolean hasWireType() {
        return switch (this) {
            case INT32, INT64, EXT_INT32, EXT_INT64 -> true;
            default -> false;
        };
    }

    /** header不属于普通意义上的容器 */
    public boolean isContainer() {
        return this == OBJECT || this == ARRAY;
    }

    /** DsonType是否是容器类型或Header */
    public boolean isContainerOrHeader() {
        return this == OBJECT || this == ARRAY || this == HEADER;
    }

    /** Dson是否是KV结构 */
    public boolean isObjectLike() {
        return this == OBJECT || this == HEADER;
    }

    public static DsonType forNumber(int number) {
        return LOOK_UP[number];
    }

}